index.js

```diff
@@ -203,12 +207,54 @@ ShareDbMongo.prototype.close = function(callback) {
 // **** Commit methods
 
 ShareDbMongo.prototype.commit = function(collectionName, id, op, snapshot, options, callback) {
+  options = options || {};
   var self = this;
   var request = createRequestForMiddleware(options, collectionName, op);
-  this._writeOp(collectionName, id, op, snapshot, function(err, result) {
+
+  var cb = callback;
+  callback = function(error, succeeded) {
+    if (error && error.code === ShareDbMongo.sessionNotStartedError().code) {
+      // Session starting is handled automatically by ShareDB, so if a session hasn't
+      // started, it actually means that a session was closed early, possibly to avoid
+      // a transaction deadlock. In this case, swallow the error and tell ShareDB that
+      // the commit hasn't succeeded, which will trigger a retry (including restarting
+      // the transaction)
+      error = null;
+      succeeded = false;
+    }
+    cb(error, succeeded);
+  };
+
+  // MongoDB locks collections when writing transactions. This can lead to deadlock if
+  // we have two different transactions attempting to update the same collection.
+  // Here, we manually keep track of the collections involved in transactions, so that
+  // we can avoid this deadlock.
+  var isLocked = this._lockedCollections[collectionName] &&
+    this._lockedCollections[collectionName] !== options.transaction;
+
+  if (isLocked) {
+    // In order to avoid deadlock, let's impose a priority system:
+    // - commits without transactions take highest priority, since they're most likely to succeed
+    // - after that, determine priority on alphabetical order of Transaction ID
+    var hasLockPriority = !options.transaction || options.transactionId < this._lockedCollections[collectionName];
+    // If we don't have priority, return a failed write to allow a retry, which should hopefully
+    // allow the other transaction to complete
+    if (!hasLockPriority) return callback(null, false);
+
+    // If we do have lock priority, we should abort the other transaction to unblock our
+    // current transaction
+    // TODO: Wait for this?
+    this.abortTransaction(this._lockedCollections[collectionName], function(error) {
+      // TODO: Handle errors
```

Comment:
I'm not really sure what to do with this error, and this feels like a bit of a code smell in the architecture of transactions.